python312Packages.vllm: 0.8.3 -> 0.9.0.1

conroy-cheers opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Based on work started in #409369.

## Things done
- Update vLLM to [v0.9.0.1](https://github.com/vllm-project/vllm/releases/tag/v0.9.0.1)
- python312Packages.bitsandbytes: 0.45.1 -> 0.46.0 (commit cherry-picked from #414382)
- Added `vllm-flash-attn` source derivation to `vllm`'s `passthru` and inputs to allow patching from an overlay

Have built and confirmed working on x86_64-linux with CUDA.

Builds and runs on aarch64-linux CUDA (Jetson Orin); the FlashAttention backend seems to have some missing kernels for sm87 but this is present before this PR already. Other backends are working OK.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

conroy-cheers committed:
python312Packages.bitsandbytes: 0.45.1 -> 0.46.0

GaetanLepage commented on pkgs/development/python-modules/vllm/default.nix:
```diff
@@ -119,36 +127,42 @@ let
+    # vllm-flash-attn normally relies on `git submodule update` to fetch cutlass and composable_kernel
+    buildPhase =
+      ''
+        ls -al
```
This was meant for debugging. You may remove it now.

GaetanLepage reviewed (commented):
Thanks for taking the work :)

GaetanLepage commented:
Can you add me as a co-author of the commits please?
```
Co-authored-by: Gaetan Lepage <gaetan@glepage.com>
```

conroy-cheers committed:
python312Packages.vllm: 0.8.3 -> 0.9.0.1

Co-authored-by: Gaetan Lepage <gaetan@glepage.com>

conroy-cheers commented:
> Can you add me as a co-author of the commits please?
> 
> ```
> Co-authored-by: Gaetan Lepage <gaetan@glepage.com>
> ```

Absolutely! Have just updated

GaetanLepage commented:
Why is `python3Packages.vllm` markes as broken on all platforms?

conroy-cheers commented:
> Why is `python3Packages.vllm` markes as broken on all platforms?

Wait, yes, I've added `flashinfer` as an input but it's only supported with CUDA. I'll fix

conroy-cheers commented:
Moved `flashinfer` to the CUDA-only dependencies and added a patch to drop the unpackaged Intel requirements, as they are optional. Tested working on CPU - wow, it builds fast without all the CUDA kernels!

GaetanLepage reviewed (approved):
Thanks!

conroy-cheers commented:
> Stumbled across this, wondering why the `intel-openmp` and `intel_extension_for_pytorch` were patched out? They're attempted to be imported upon e.g. `vllm serve` when running CPU-only on my machine, causing a crash since it's not installed

Those dependencies weren't (/ aren't?) available in Nixpkgs. FYI I'm pretty sure CPU inference is working as of [0.9.1](https://github.com/NixOS/nixpkgs/pull/416385)
